swbemservices.go

```diff
@@ -0,0 +1,287 @@
+	s.queries = nil //set channel to nil so we know it is closed
+	s.sWbemLocatorIDispatch.Release()
+	s.sWbemLocatorIUnknown.Release()
+	ole.CoUninitialize()
+	runtime.UnlockOSThread()
+	fmt.Println("closeBackground: Finished")
+	return nil
+}
+
+func (s *SWbemServices) process(initError chan error) {
+	fmt.Println("process: starting background thread initialization")
+	//All OLE/WMI calls must happen on the same initialized thead, so lock this goroutine
+	runtime.LockOSThread()
+	//defer s.closeBackground() //Does this run on the same locked thread? Can't rely on s.closing as select hasn't started yet
+
+	err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
+	if err != nil {
+		oleCode := err.(*ole.OleError).Code()
+		if oleCode != ole.S_OK && oleCode != S_FALSE {
+			initError <- fmt.Errorf("ole.CoInitializeEx error: %v", err)
+			return
+		}
+	}
+
+	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
+	if err != nil {
+		initError <- fmt.Errorf("CreateObject SWbemLocator error: %v", err)
+		return
+	} else if unknown == nil {
+		initError <- ErrNilCreateObject
+		return
+	}
+	s.sWbemLocatorIUnknown = unknown
+
+	dispatch, err := s.sWbemLocatorIUnknown.QueryInterface(ole.IID_IDispatch)
+	if err != nil {
+		initError <- fmt.Errorf("SWbemLocator QueryInterface error: %v", err)
+		return
+	}
+	s.sWbemLocatorIDispatch = dispatch
+
+	// we can't do the ConnectServer call outside the loop unless we find a way to track and re-init the connectServerArgs
+	fmt.Println("process: initialized. closing initError")
+	close(initError)
+	fmt.Println("process: waiting for queries")
+	for {
+		select {
+		case q, ok := <-s.queries:
+			if ok {
+				fmt.Printf("process: new query: len(query)=%d\n", len(q.query))
+				errQuery := s.queryBackground(q)
+				fmt.Println("process: s.queryBackground finished")
+				if errQuery != nil {
+					q.finished <- errQuery
+				}
+				close(q.finished)
+			} else {
+				fmt.Println("process: queries channel closed")
+				errClose := s.closeBackground()
+				if errClose == nil && s.closeError != nil {
+					s.closeError <- errClose
+				}
+				if s.closeError != nil {
+					close(s.closeError)
+				}
+			}
+		default:
+		}
+	}
+}
+
+// Query runs the WQL query and appends the values to dst.
+//
+// dst must have type *[]S or *[]*S, for some struct type S. Fields selected in
+// the query must have the same name in dst. Supported types are all signed and
+// unsigned integers, time.Time, string, bool, or a pointer to one of those.
+// Array types are not supported.
+//
+// By default, the local machine and default namespace are used. These can be
+// changed using connectServerArgs. See
+// http://msdn.microsoft.com/en-us/library/aa393720.aspx for details.
+func (s *SWbemServices) Query(query string, dst interface{}, connectServerArgs ...interface{}) error {
+	if s == nil || s.sWbemLocatorIDispatch == nil {
+		return fmt.Errorf("SWbemServices is not Initialized")
+	}
+	if s.queries == nil {
```

Comment:
Possible race condition here. If it closes between the nil check and putting stuff on the channel. I'd make a single lock and you need to acquire it to insert a query or close. Since everything queues up anyway, it shouldn't be a problem.